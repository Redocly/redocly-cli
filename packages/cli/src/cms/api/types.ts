export type ProjectSourceResponse = {
  branchName: string;
  contentPath: string;
  isInternal: boolean;
};

export type UpsertRemoteResponse = {
  id: string;
  type: 'CICD';
  mountPath: string;
  mountBranchName: string;
  organizationId: string;
  projectId: string;
};

export type ListRemotesResponse = {
  object: 'list';
  page: {
    endCursor: string;
    startCursor: string;
    haxNextPage: boolean;
    hasPrevPage: boolean;
    limit: number;
    total: number;
  };
  items: Remote[];
};

export type Remote = {
  mountPath: string;
  type: string;
  autoSync: boolean;
  autoMerge: boolean;
  createdAt: string;
  updatedAt: string;
  providerType: string;
  namespaceId: string;
  repositoryId: string;
  projectId: string;
  mountBranchName: string;
  contentPath: string;
  credentialId: string;
  branchName: string;
  contentType: string;
  id: string;
};

export type PushResponse = {
  id: string;
  remoteId: string;
  isMainBranch: boolean;
  isOutdated: boolean;
  hasChanges: boolean;
  replace: boolean;
  scoutJobId: string | null;
  uploadedFiles: Array<{ path: string; mimeType: string }>;
  commit: {
    branchName: string;
    message: string;
    createdAt: string | null;
    namespaceId: string | null;
    repositoryId: string | null;
    url: string | null;
    sha: string | null;
    author: {
      name: string;
      email: string;
      image: string | null;
    };
    statuses: Array<{
      name: string;
      description: string;
      status: 'pending' | 'running' | 'success' | 'failed';
      url: string | null;
    }>;
  };
  remote: {
    commits: {
      sha: string;
      branchName: string;
    }[];
  };
  status: PushStatusResponse;
};

export type DeploymentStatusResponse = {
  deploy: {
    url: string | null;
    status: DeploymentStatus;
  };
  scorecard: ScorecardItem[];
};

export type PushStatusResponse = {
  preview: DeploymentStatusResponse;
  production: DeploymentStatusResponse;
};

export type ScorecardItem = {
  name: string;
  status: PushStatusBase;
  description: string;
  url: string;
};

export type PushStatusBase = 'pending' | 'success' | 'running' | 'failed';

export type DeploymentStatus = 'skipped' | PushStatusBase;
