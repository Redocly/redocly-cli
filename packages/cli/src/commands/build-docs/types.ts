export type BuildDocsOptions = {
  watch?: boolean;
  cdn?: boolean;
  output?: string;
  title?: string;
  disableGoogleFont?: boolean;
  port?: number;
  templateFileName?: string;
  templateOptions?: any;
  redocOptions?: any;
  redocCurrentVersion: string;
};

export type BuildDocsArgv = {
  api: string;
  o: string;
  cdn: boolean;
  title?: string;
  disableGoogleFont?: boolean;
  template?: string;
  templateOptions: Record<string, any>;
  theme: {
    openapi: string | Record<string, unknown>;
  };
  config?: string;
};
