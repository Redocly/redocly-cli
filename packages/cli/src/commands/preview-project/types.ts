import { PRODUCT_PACKAGES, PRODUCT_PLANS } from './constants';

export type Product = keyof typeof PRODUCT_PACKAGES;
export type ProductPlan = typeof PRODUCT_PLANS[number];

export type PreviewProjectOptions = {
  product?: Product | string;
  plan: ProductPlan | string;
  port?: number;
  'source-dir': string;
  config?: string;
};
