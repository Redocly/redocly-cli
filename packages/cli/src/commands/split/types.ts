import {
  Oas3Schema,
  Oas3_1Schema,
  Oas3Definition,
  Oas3_1Definition,
  Oas3Components,
  Oas3PathItem,
  Oas3Paths,
  Oas3ComponentName,
  Oas3_1Webhooks,
  Oas2Definition,
  Referenced
} from "@redocly/openapi-core";
export { Oas3_1Definition, Oas3Definition, Oas2Definition, Oas3Components, Oas3Paths, Oas3PathItem, Oas3ComponentName, Oas3_1Schema, Oas3Schema, Oas3_1Webhooks, Referenced }
export type Definition = Oas3_1Definition | Oas3Definition | Oas2Definition;
export interface ComponentsFiles {
  [schemas: string]: any;
}
export interface refObj {
  [$ref: string]: string;
}

export const COMPONENTS = 'components';
export const PATHS = 'paths';
export const WEBHOOKS = 'webhooks';
export const xWEBHOOKS = 'x-webhooks';
export const componentsPath = `#/${COMPONENTS}/`;

enum OPENAPI3_METHOD {
  Get = 'get',
  Put = 'put',
  Post = 'post',
  Delete = 'delete',
  Options = 'options',
  Head = 'head',
  Patch = 'patch',
  Trace = 'trace'
}

export const OPENAPI3_METHOD_NAMES: OPENAPI3_METHOD[] = [
  OPENAPI3_METHOD.Get,
  OPENAPI3_METHOD.Put,
  OPENAPI3_METHOD.Post,
  OPENAPI3_METHOD.Delete,
  OPENAPI3_METHOD.Options,
  OPENAPI3_METHOD.Head,
  OPENAPI3_METHOD.Patch,
  OPENAPI3_METHOD.Trace
];

export enum OPENAPI3_COMPONENT {
  Schemas = 'schemas',
  Responses = 'responses',
  Parameters = 'parameters',
  Examples = 'examples',
  Headers = 'headers',
  RequestBodies = 'requestBodies',
  Links = 'links',
  Callbacks = 'callbacks',
  SecuritySchemes = 'securitySchemes'
}

export const OPENAPI3_COMPONENT_NAMES: OPENAPI3_COMPONENT[] = [
  OPENAPI3_COMPONENT.RequestBodies,
  OPENAPI3_COMPONENT.Schemas,
  OPENAPI3_COMPONENT.Responses,
  OPENAPI3_COMPONENT.Parameters,
  OPENAPI3_COMPONENT.Examples,
  OPENAPI3_COMPONENT.Headers,
  OPENAPI3_COMPONENT.Links,
  OPENAPI3_COMPONENT.Callbacks,
  OPENAPI3_COMPONENT.SecuritySchemes
];
