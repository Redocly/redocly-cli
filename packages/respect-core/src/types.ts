import type { FromSchema } from 'json-schema-to-ts';
import type {
  parameter,
  operationMethod,
  sourceDescriptionSchema,
  infoObject,
  requestBody,
  replacement,
  criteriaObject,
  step,
  workflow,
  reusableObject,
  onSuccessObject,
  onFailureObject,
  extendedOperation,
  extendedSecurity,
} from './arazzo-schema.js';
import type { Faker } from './modules/faker.js';
import type { OperationDetails } from './modules/description-parser/index.js';
import type { ApiFetcher } from './utils/api-fetcher.js';
import type { RespectOptions } from './handlers/run.js';
import type { Config, CollectFn, RuleSeverity } from '@redocly/openapi-core';

export type OperationMethod = FromSchema<typeof operationMethod>;
export type ResponseContext = {
  statusCode: number;
  body: any;
  header: Record<string, string>;
  contentType?: string;
  time?: number;
} & Record<string, any>;
export type SourceDescription = FromSchema<typeof sourceDescriptionSchema>;
type ArazzoParameter = FromSchema<typeof parameter>;
export type InfoObject = FromSchema<typeof infoObject>;
export type RequestBody = FromSchema<typeof requestBody>;
export type Replacement = FromSchema<typeof replacement>;
export type CriteriaObject = FromSchema<typeof criteriaObject>;
export type ReusableObject = FromSchema<typeof reusableObject>;
export type OnSuccessObject = FromSchema<typeof onSuccessObject>;
export type OnFailureObject = FromSchema<typeof onFailureObject>;
export type ExtendedOperation = FromSchema<typeof extendedOperation>;
export type ExtendedSecurity = FromSchema<typeof extendedSecurity>;
type ArazzoStep = FromSchema<typeof step>;
type ArazzoWorkflow = FromSchema<typeof workflow> & {
  steps: Step[];
};

export type AdditionalParameterProperties = {
  style?: string;
  target?: string;
  required?: boolean;
  schema?: Record<string, any>;
  example?: unknown;
  examples?: Record<string, any> | unknown;
};
type ExtendedParameter<T> = T & AdditionalParameterProperties;
export type Parameter = ExtendedParameter<ArazzoParameter>;
export type ResolvedParameter = Parameter & {
  in?: 'header' | 'query' | 'path' | 'cookie';
  name: string;
};
export type VerboseLog = {
  method: OperationMethod;
  path: string;
  host: string;
  body?: any;
  headerParams?: Record<string, string>;
  statusCode?: number;
  responseTime?: number;
};
type AdditionalStepProps = {
  verboseLog?: VerboseLog;
  response: ResponseContext;
  checks: Check[];
  retriesLeft?: number;
};

export type Step = ArazzoStep & AdditionalStepProps;
export type Workflow = Omit<ArazzoWorkflow, 'steps'> & { steps: Step[]; time?: number };
export type RunArgv = Omit<RespectOptions, 'files'> & {
  file: string;
  testDescription?: TestDescription;
  input?: string | string[];
  server?: string | string[];
  severity?: string | string[];
};

export type CommandArgs<T> = {
  argv: T;
  config: Config;
  version: string;
  collectSpecData?: CollectFn;
};

export interface RequestContext {
  body: any;
  header: Record<string, string | number | boolean>;
  path?: Record<string, string | number | boolean>;
  method?: string;
  url?: string;
  query?: Record<string, string | number | boolean>;
}

export type ParsedParameters = {
  queryParams: Record<string, string>;
  pathParams: Record<string, string | number>;
  headerParams: Record<string, string>;
};
export type AppOptions = {
  workflowPath: string;
  workflow?: string | string[];
  skip?: string | string[];
  verbose?: boolean;
  harOutput?: string;
  jsonOutput?: string;
  metadata?: Record<string, any>;
  input?: string | string[];
  server?: string | string[];
  severity?: string | string[];
  mutualTls?: Partial<TestContext['mtlsCerts']>;
};
export type RegexpSuccessCriteria = {
  condition: string;
  context: string;
  type: 'regex';
};
export type JsonPathVerison = 'draft-goessner-dispatch-jsonpath-00';

export type JsonPathSuccessCriterionObject = {
  type: 'jsonpath';
  version: JsonPathVerison;
};

export type JsonPathSuccessCriteria = {
  condition: string;
  context: string;
  type: 'jsonpath' | JsonPathSuccessCriterionObject;
};
export type Ref = {
  $ref: string;
};
export type PublicWorkflow = {
  outputs?: Record<string, any>;
  inputs?: Record<string, any>;
  steps: Record<string, PublicWorkflowStep>;
};
export type PublicStep = {
  outputs?: Record<string, any>;
};

export type PublicWorkflowStep = {
  outputs?: Record<string, any>;
  request?: RequestContext;
  response?: ResponseContext;
};

export interface InputSchema {
  type: string;
  properties?: {
    [key: string]: any;
  };
  format?: string;
}

export type StepInnerContext = {
  $response: ResponseContext | undefined;
  $request: RequestContext | undefined;
  $outputs: Record<string, any>;
};
export type WorkflowInnerContext = {
  $steps: Record<string, PublicWorkflowStep>;
  $outputs: Record<string, any>;
};
export type RuntimeExpressionContext = {
  $workflows: Record<string, PublicWorkflow>;
  $sourceDescriptions: Record<string, any>;
  $faker: Faker;
  $steps: Record<string, PublicStep>;
  $response?: Partial<ResponseContext>;
  $request?: Partial<RequestContext>;
  $outputs?: Record<string, any>;
  $inputs?: Record<string, any>;
  $components?: Record<string, any>;
  $url?: string;
  $method?: string;
  $statusCode?: number;
};

export type RunWorkflowInput = {
  workflowInput: Workflow | string;
  ctx: TestContext;
  fromStepId?: string;
  parentStepId?: string;
  skipLineSeparator?: boolean;
  invocationContext?: string;
};

export type ArrazoItemExecutionResult = StepExecutionResult | WorkflowExecutionResult;

export type ExecutionStatus = 'success' | 'error' | 'warn';
export interface StepExecutionResult {
  type: 'step';
  stepId: string;
  workflowId: string;
  totalTimeMs: number;

  retriesLeft?: number; // number of retries
  status: ExecutionStatus;

  invocationContext?: string;

  request?: {
    method: string;
    url: string;
    headers: Record<string, string | number | boolean>;
    body: any;
  };
  response?: {
    statusCode: number;
    body: any;
    headers: Record<string, string | number | boolean>;
    time: number;
  };
  checks: (Check & { status: ExecutionStatus })[];
}

export interface WorkflowExecutionResult {
  type: 'workflow';
  workflowId: string;
  stepId?: string; // for child workflows
  sourceDescriptionName?: string; // maybe drop for now

  startTime: number;
  endTime: number;
  totalTimeMs: number;

  executedSteps: (Step | WorkflowExecutionResult)[];
  invocationContext?: string;
  ctx: TestContext;
  globalTimeoutError: boolean;
}

export type WorkflowExecutionResultJson = Omit<WorkflowExecutionResult, 'executedSteps' | 'ctx'> & {
  executedSteps: (StepExecutionResult | WorkflowExecutionResultJson)[];
  status: ExecutionStatus;
  totalRequests: number;
  globalTimeoutError: boolean;
};

export type TestContext = RuntimeExpressionContext & {
  executedSteps: (Step | WorkflowExecutionResult)[];
  arazzo: string;
  info: InfoObject & Record<string, any>;
  sourceDescriptions?: SourceDescription[];
  workflows: Workflow[];
  options: AppOptions;
  testDescription: TestDescription;
  harLogs: any;
  components?: Record<string, any>;
  secretFields: Set<string>;
  severity: Record<string, RuleSeverity>;
  mtlsCerts?: {
    clientCert?: string;
    clientKey?: string;
    caCert?: string;
  };
  apiClient: ApiFetcher;
};

export type TestDescription = Partial<
  Pick<
    TestContext & WorkflowInnerContext & StepInnerContext,
    'workflows' | 'arazzo' | 'info' | 'sourceDescriptions' | '$outputs' | 'components'
  >
>;

export type Check = {
  severity: RuleSeverity;
  passed: boolean;
  name: string;
  message?: string;
  condition?: string;
};

export interface ResultsOfTests {
  passed: number;
  failed: number;
  total: number;
  warnings: number;
  skipped: number;
}

export type CalculatedResults = {
  workflows: ResultsOfTests;
  steps: ResultsOfTests;
  checks: ResultsOfTests;
  totalRequests: number;
};
export type StepCallContext = {
  $response?: ResponseContext;
  $request?: RequestContext;
  $inputs?: Record<string, any>;
};

export type JsonLogs = {
  files: Record<
    string,
    {
      totalRequests: number;
      executedWorkflows: WorkflowExecutionResultJson[];
    }
  >;
  status: string;
  totalTime: number;
  globalTimeoutError: boolean;
};

export type DescriptionChecks = {
  checks: Check[];
  descriptionOperation: OperationDetails;
  $response: ResponseContext;
};
