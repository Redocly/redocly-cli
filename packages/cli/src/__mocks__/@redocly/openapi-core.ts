import { ConfigFixture } from './../../__tests__/fixtures/config';
import { Document, ResolveError } from '@redocly/openapi-core';
import { firstDocument, secondDocument } from '../documents';

export const __redoclyClient = {
  isAuthorizedWithRedocly: jest.fn().mockResolvedValue(true),
  isAuthorizedWithRedoclyByRegion: jest.fn().mockResolvedValue(true),
  login: jest.fn(),
  registryApi: {
    setAccessTokens: jest.fn(),
    authStatus: jest.fn(),
    prepareFileUpload: jest.fn().mockResolvedValue({
      signedUploadUrl: 'signedUploadUrl',
      filePath: 'filePath',
    }),
    pushApi: jest.fn(),
  },
};

export const RedoclyClient = jest.fn(() => __redoclyClient);
export const loadConfig = jest.fn(() => ConfigFixture);
export const getMergedConfig = jest.fn();
export const lint = jest.fn();
export const bundle = jest.fn(() => ({ bundle: { parsed: null }, problems: null }));
export const getTotals = jest.fn(() => ({ errors: 0 }));
export const formatProblems = jest.fn();
export const slash = jest.fn();
export const findConfig = jest.fn();
export const doesYamlFileExist = jest.fn();
export const bundleDocument = jest.fn(() => Promise.resolve({ problems: {} }));
export const detectOpenAPI = jest.fn();
export const isAbsoluteUrl = jest.fn();

export class BaseResolver {
  cache = new Map<string, Promise<Document | ResolveError>>();

  getFiles = jest.fn();
  resolveExternalRef = jest.fn();
  loadExternalRef = jest.fn;
  parseDocument = jest.fn();
  resolveDocument = jest
    .fn()
    .mockImplementationOnce(() =>
      Promise.resolve({ source: { absoluteRef: 'ref' }, parsed: firstDocument })
    )
    .mockImplementationOnce(() =>
      Promise.resolve({ source: { absoluteRef: 'ref' }, parsed: secondDocument })
    );
}

export enum OasVersion {
  Version2 = 'oas2',
  Version3_0 = 'oas3_0',
  Version3_1 = 'oas3_1',
}

export enum Oas3Operations {
  get = 'get',
  put = 'put',
  post = 'post',
  delete = 'delete',
  options = 'options',
  head = 'head',
  patch = 'patch',
  trace = 'trace',
}
