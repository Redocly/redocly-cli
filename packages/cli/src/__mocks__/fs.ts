export const existsSync = jest.fn();
export const readFileSync = jest.fn(() => '');
export const statSync = jest.fn(() => ({ size: 0 }));
export const createReadStream = jest.fn();
export const writeFileSync = jest.fn();
export const mkdirSync = jest.fn();
