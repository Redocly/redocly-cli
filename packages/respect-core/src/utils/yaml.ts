// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore broken types for js-yaml
import { JSON_SCHEMA, types, load, dump } from 'js-yaml';
import { readFileSync } from 'fs';

import type { LoadOptions, DumpOptions } from 'js-yaml';

const DEFAULT_SCHEMA_WITHOUT_TIMESTAMP = JSON_SCHEMA.extend({
  implicit: [types.merge],
  explicit: [types.binary, types.omap, types.pairs, types.set],
});

export function parseYaml(str: string, opts?: LoadOptions): unknown {
  return load(str, { schema: DEFAULT_SCHEMA_WITHOUT_TIMESTAMP, ...opts });
}

export function stringifyYaml(obj: any, opts?: DumpOptions): string {
  return dump(obj, opts);
}

export function readYaml(path: string): unknown {
  return parseYaml(readFileSync(path, 'utf-8'));
}
