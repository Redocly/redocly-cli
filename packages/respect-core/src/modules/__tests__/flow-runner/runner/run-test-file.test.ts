import { makeDocumentFromString, lint, bundle, type LocationObject } from '@redocly/openapi-core';
import * as fs from 'node:fs';
import { type Step, type TestContext } from '../../../../types.js';
import { runTestFile, runStep } from '../../../flow-runner/index.js';
import { readYaml } from '../../../../utils/yaml.js';

vi.mock('../../../../utils/yaml.js', () => {
  const originalModule = vi.importActual('../../../../utils/yaml.js');

  return {
    ...originalModule, // In case there are other exports you want to preserve
    readYaml: vi.fn(),
  };
});

vi.mock('@redocly/openapi-core', async () => {
  const originalModule = await vi.importActual('@redocly/openapi-core');

  return {
    ...originalModule, // Preserve other exports
    formatProblems: vi.fn(), // Mock formatProblems to do nothing
    lint: vi.fn(),
    bundle: vi.fn(),
  };
});

vi.mock('../../../flow-runner/run-step.js', () => ({
  runStep: vi.fn(),
}));

vi.mock('node:fs', () => {
  const actual = vi.importActual('node:fs');
  const mockExistsSync = vi.fn();

  return {
    __esModule: true,
    default: {
      ...actual,
      existsSync: mockExistsSync,
    },
    existsSync: mockExistsSync,
  };
});

const mockExistsSync = vi.mocked(fs.existsSync);

describe('runTestFile', () => {
  beforeEach(() => {
    mockExistsSync.mockImplementation((path: any) => {
      // Always return true for test files to allow readYaml to be called
      if (path.includes('test.yaml') || path.includes('test.yml')) {
        return true;
      }

      // Return true for source description files (both relative and absolute paths)
      if (path.includes('cats.yaml')) {
        return true;
      }
      // Return false for non-existing files
      if (path.includes('not-existing')) {
        return false;
      }
      return false;
    });
    vi.mocked(readYaml).mockResolvedValue({
      openapi: '1.0.0',
      info: { title: 'Cat Facts API', version: '1.0' },
    });
    vi.mocked(lint).mockResolvedValue([]);
  });

  it(`should trow error if filename is not correct`, async () => {
    await expect(runTestFile({ file: '' }, {})).rejects.toThrowError('Invalid file name');
  });

  it(`should trow error if file is not valid Arazzo test file`, async () => {
    const mockDocument = makeDocumentFromString(
      JSON.stringify({
        openapi: '1.0.0',
        info: { title: 'Cat Facts API', version: '1.0' },
      }),
      'test.yml'
    );

    vi.mocked(readYaml).mockResolvedValue(mockDocument.parsed);

    await expect(runTestFile({ file: 'test.yaml' }, {})).rejects.toThrowError(
      'No test files found. File test.yaml does not follows naming pattern "*.[yaml | yml | json]" or have not valid "Arazzo" description.'
    );
  });

  it('should throw Found errors in Arazzo description error when contains lint errors', async () => {
    const testDescription = {
      workflows: [],
    };

    const mockDocument = makeDocumentFromString(
      JSON.stringify({
        arazzo: '1.0.1',
        info: { title: 'Cat Facts API', version: '1.0' },
        workflows: [
          {
            workflowId: 'get-bird-workflow',
            parameters: [{ in: 'header', name: 'IMF-KEY', value: 'test-key' }],
          },
        ],
      }),
      'test.yml'
    );

    vi.mocked(lint).mockResolvedValueOnce([
      {
        ruleId: 'struct',
        severity: 'error',
        message: 'The field `sourceDescriptions` must be present on this level.',
        from: undefined,
        location: [{} as LocationObject],
        suggest: [],
      },
      {
        ruleId: 'struct',
        severity: 'error',
        message: 'Property `sourceDescriptionsd` is not expected here.',
        suggest: ['sourceDescriptions'],
        from: undefined,
        location: [{} as LocationObject],
      },
    ]);

    vi.mocked(readYaml).mockResolvedValue(mockDocument.parsed);

    await expect(
      runTestFile(
        {
          file: 'test.yaml',
          testDescription,
        },
        {}
      )
    ).rejects.toMatchSnapshot();
  });

  it('should call runStep once', async () => {
    const mockDocument = makeDocumentFromString(
      JSON.stringify({
        arazzo: '1.0.1',
        info: {
          title: 'Cat Facts API',
          version: '1.0',
        },
        sourceDescriptions: [
          {
            name: 'cats',
            type: 'openapi',
            url: 'api-samples/cats.yaml',
          },
        ],
        workflows: [
          {
            workflowId: 'get-bird-workflow',
            parameters: [
              {
                in: 'header',
                name: 'IMF-KEY',
                value: 'test-key',
              },
              {
                in: 'cookie',
                name: 'workflowLevelParam',
                value: 'workflowcookie',
              },
            ],
            steps: [
              {
                stepId: 'delete-step',
                'x-operation': {
                  url: 'http://localhost:3000/delete-mock',
                  method: 'delete',
                },
                successCriteria: [
                  {
                    condition: '$statusCode == 204',
                  },
                  {
                    condition: '$response.header.content-type == "text/plain;charset=UTF-8"',
                  },
                ],
              },
            ],
          },
        ],
      }),
      'api-test-framework/test.yml'
    );

    vi.mocked(readYaml).mockResolvedValue(mockDocument.parsed);
    vi.mocked(lint).mockResolvedValueOnce([]);
    vi.mocked(bundle).mockResolvedValueOnce({
      bundle: {
        parsed: mockDocument.parsed,
      },
    } as any);

    await runTestFile(
      {
        file: 'test.yaml',
      },
      {}
    );

    expect(runStep).toHaveBeenCalledTimes(1);
  });

  it('should call runStep multiple times including from dependsOn workflow', async () => {
    const mockDocument = makeDocumentFromString(
      JSON.stringify({
        arazzo: '1.0.1',
        info: {
          title: 'Cat Facts API',
          version: '1.0',
        },
        sourceDescriptions: [
          {
            name: 'cats',
            type: 'openapi',
            url: 'api-samples/cats.yaml',
          },
        ],
        workflows: [
          {
            workflowId: 'get-bird-workflow',
            parameters: [
              {
                in: 'header',
                name: 'IMF-KEY',
                value: 'test-key',
              },
              {
                in: 'cookie',
                name: 'workflowLevelParam',
                value: 'workflowcookie',
              },
            ],
            steps: [
              {
                stepId: 'delete-step',
                'x-operation': {
                  url: 'http://localhost:3000/delete-mock',
                  method: 'delete',
                },
                successCriteria: [
                  {
                    condition: '$statusCode == 204',
                  },
                ],
              },
            ],
          },
          {
            workflowId: 'second-workflow',
            parameters: [
              {
                in: 'header',
                name: 'IMF-KEY',
                value: 'test-key',
              },
            ],
            dependsOn: ['get-bird-workflow'],
            steps: [
              {
                stepId: 'delete-mock',
                'x-operation': {
                  url: 'http://localhost:3000/delete-mock',
                  method: 'delete',
                },
                successCriteria: [
                  {
                    condition: '$statusCode == 204',
                  },
                ],
              },
            ],
          },
        ],
      }),
      'test.yml'
    );

    vi.mocked(readYaml).mockResolvedValue(mockDocument.parsed);
    vi.mocked(lint).mockResolvedValueOnce([]);
    vi.mocked(bundle).mockResolvedValueOnce({
      bundle: {
        parsed: mockDocument.parsed,
      },
    } as any);

    await runTestFile(
      {
        file: 'test.yaml',
      },
      {}
    );

    // called 3 times, one for each step from each workflow and one from dependsOn
    expect(runStep).toHaveBeenCalledTimes(3);
  }, 8000);

  it('should throw an error when dependsOn has not existing workflowId', async () => {
    const mockDocument = makeDocumentFromString(
      JSON.stringify({
        arazzo: '1.0.1',
        info: {
          title: 'Cat Facts API',
          version: '1.0',
        },
        sourceDescriptions: [
          {
            name: 'cats',
            type: 'openapi',
            url: 'api-samples/cats.yaml',
          },
        ],
        workflows: [
          {
            workflowId: 'get-bird-workflow',
            parameters: [
              {
                in: 'header',
                name: 'IMF-KEY',
                value: 'test-key',
              },
              {
                in: 'cookie',
                name: 'workflowLevelParam',
                value: 'workflowcookie',
              },
            ],
            steps: [
              {
                stepId: 'delete-step',
                'x-operation': {
                  url: 'http://localhost:3000/delete-mock',
                  method: 'delete',
                },
                successCriteria: [
                  {
                    condition: '$statusCode == 204',
                  },
                ],
              },
            ],
          },
          {
            workflowId: 'second-workflow',
            parameters: [
              {
                in: 'header',
                name: 'IMF-KEY',
                value: 'test-key',
              },
            ],
            dependsOn: ['get-bird-workflow', 'not-existing-workflowId'],
            steps: [
              {
                stepId: 'delete-mock',
                'x-operation': {
                  url: 'http://localhost:3000/delete-mock',
                  method: 'delete',
                },
                successCriteria: [
                  {
                    condition: '$statusCode == 204',
                  },
                ],
              },
            ],
          },
        ],
      }),
      'api-test-framework/test.yml'
    );

    vi.mocked(readYaml).mockResolvedValue(mockDocument.parsed);
    vi.mocked(lint).mockResolvedValueOnce([]);
    vi.mocked(bundle).mockResolvedValueOnce({
      bundle: {
        parsed: mockDocument.parsed,
      },
    } as any);

    await expect(runTestFile({ file: 'test.yaml' }, {})).rejects.toThrow(
      expect.objectContaining({
        // @ts-ignore
        message: expect.stringContaining('Workflow', 'not-existing-workflowId', 'not found'),
      })
    );
  }, 8000);

  it('should throw an error when dependsOn has workflowId with not successful steps expectations', async () => {
    const mockDocument = makeDocumentFromString(
      JSON.stringify({
        arazzo: '1.0.1',
        info: {
          title: 'Cat Facts API',
          version: '1.0',
        },
        sourceDescriptions: [
          {
            name: 'cats',
            type: 'openapi',
            url: 'api-samples/cats.yaml',
          },
        ],
        workflows: [
          {
            workflowId: 'get-bird-workflow',
            parameters: [
              {
                in: 'header',
                name: 'IMF-KEY',
                value: 'test-key',
              },
              {
                in: 'cookie',
                name: 'workflowLevelParam',
                value: 'workflowcookie',
              },
            ],
            steps: [
              {
                stepId: 'delete-step',
                'x-operation': {
                  url: 'http://localhost:3000/delete-mock',
                  method: 'delete',
                },
                successCriteria: [
                  {
                    condition: '$statusCode == 204',
                  },
                ],
              },
            ],
          },
          {
            workflowId: 'second-workflow',
            dependsOn: ['get-bird-workflow'],
            parameters: [
              {
                in: 'header',
                name: 'IMF-KEY',
                value: 'test-key',
              },
            ],
            steps: [
              {
                stepId: 'delete-mock',
                'x-operation': {
                  url: 'http://localhost:3000/delete-mock',
                  method: 'delete',
                },
                successCriteria: [
                  {
                    condition: '$statusCode == 204',
                  },
                ],
              },
            ],
          },
        ],
      }),
      'api-test-framework/test.yml'
    );

    vi.mocked(readYaml).mockResolvedValue(mockDocument.parsed);
    vi.mocked(lint).mockResolvedValueOnce([]);
    vi.mocked(bundle).mockResolvedValueOnce({
      bundle: {
        parsed: mockDocument.parsed,
      },
    } as any);

    vi.mocked(runStep).mockImplementation(
      async ({ step, ctx }: { step: Step; ctx: TestContext }) => {
        step.checks = [{ name: step.stepId, passed: false, severity: 'error' }];
        ctx.executedSteps.push(step);
      }
    );

    await expect(
      runTestFile(
        {
          file: 'test.yaml',
        },
        {}
      )
    ).rejects.toThrowError('Dependent workflows has failed steps');
  }, 8000);

  it('should throw an error when sourcedescription OpenAPI file does not exist', async () => {
    const mockDocument = makeDocumentFromString(
      JSON.stringify({
        arazzo: '1.0.1',
        info: {
          title: 'Cat Facts API',
          version: '1.0',
        },
        sourceDescriptions: [
          {
            name: 'cats',
            type: 'openapi',
            url: 'api-samples/not-existing.yaml',
          },
        ],
        workflows: [
          {
            workflowId: 'get-bird-workflow',
            steps: [
              {
                stepId: 'delete-step',
                'x-operation': {
                  url: 'http://localhost:3000/delete-mock',
                  method: 'delete',
                },
                successCriteria: [
                  {
                    condition: '$statusCode == 204',
                  },
                ],
              },
            ],
          },
        ],
      }),
      'api-test-framework/test.yml'
    );

    vi.mocked(readYaml).mockResolvedValue(mockDocument.parsed);
    vi.mocked(lint).mockResolvedValueOnce([]);
    vi.mocked(bundle).mockResolvedValueOnce({
      bundle: {
        parsed: mockDocument.parsed,
      },
    } as any);
    await expect(runTestFile({ file: 'test.yaml' }, {})).rejects.toThrowError(
      `Could not find source description file 'api-samples/not-existing.yaml' at path 'test.yaml'`
    );
  });

  it('should throw an error when sourcedescription Arazzo file does not exist', async () => {
    const mockDocument = makeDocumentFromString(
      JSON.stringify({
        arazzo: '1.0.1',
        info: {
          title: 'Cat Facts API',
          version: '1.0',
        },
        sourceDescriptions: [
          {
            name: 'cats',
            type: 'openapi',
            url: 'api-samples/cats.yaml',
          },
          {
            name: 'cats',
            type: 'arazzo',
            url: 'api-samples/not-existing-arazzo.yaml',
          },
        ],
        workflows: [
          {
            workflowId: 'get-bird-workflow',
            steps: [
              {
                stepId: 'delete-step',
                'x-operation': {
                  url: 'http://localhost:3000/delete-mock',
                  method: 'delete',
                },
                successCriteria: [
                  {
                    condition: '$statusCode == 204',
                  },
                ],
              },
            ],
          },
        ],
      }),
      'api-test-framework/test.yml'
    );

    vi.mocked(readYaml).mockResolvedValue(mockDocument.parsed);
    vi.mocked(lint).mockResolvedValueOnce([]);
    vi.mocked(bundle).mockResolvedValueOnce({
      bundle: {
        parsed: mockDocument.parsed,
      },
    } as any);
    await expect(runTestFile({ file: 'test.yaml' }, {})).rejects.toThrowError(
      `Could not find source description file 'not-existing-arazzo.yaml' at path 'api-samples/not-existing-arazzo.yaml'`
    );
  });
});
