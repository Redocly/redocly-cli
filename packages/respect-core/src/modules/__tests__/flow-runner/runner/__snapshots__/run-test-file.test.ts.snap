// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runTestFile > should throw Found errors in Arazzo description error when contains lint errors 1`] = `[Error: [31mFound errors in Arazzo description[39m [1mtest.yaml[22m]`;
